### PR TITLE
[WIP] Implement VK_NV_device_generated_commands (2)

### PR DESCRIPTION
**Pull request by pixeljetstream**
re-opened version of original #1637 

unit tests are yet to be written
could attempt to split this up in multiple merges if desired

it would be good to have feedback on the general approach taken, specifically the refactoring of the graphics pipeline validation

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -1456,9 +1469,30 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
         }
     }
 
-    if (ValidateGraphicsPipelineShaderState(pPipeline)) {
-        skip = true;
+    auto pCreateInfo = pPipeline->graphicsPipelineCI.ptr();
```
Comment (mark-lunarg) [inline layers/core_validation.cpp:1472]: All local variables need to be lower_snake_case, please -- I think the code checker might even call these not, not sure if that's in yet though. We apologize in advance if you were following the surrounding style -- progress has recently been made to scrub the source but haven't quite finished.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -1652,28 +1686,103 @@ bool CoreChecks::ValidatePipelineUnlocked(const PIPELINE_STATE *pPipeline, uint3
...
+        const VkGraphicsPipelineCreateInfo *pA = pPipeline->graphicsPipelineCI.ptr();
+        for (size_t i = 0; i < pPipeline->shader_group_references.size(); i++) {
+            skip |= ValidatePipelineReferences(pPipeline->shader_group_references[i].pipeline_state->pipeline,
+                                               "VUID-VkGraphicsPipelineShaderGroupsCreateInfoNV-pPipelines-parameter",
+                                               "vkCreateGraphicsPipelines()");
+
+            const PIPELINE_STATE *pReference = pPipeline->shader_group_references[i].pipeline_state.get();
+            const VkGraphicsPipelineCreateInfo *pB = pReference->graphicsPipelineCI.ptr();
+
+            // FIXME just a basic test, full validity would need to do a deep comparison of all pNext chains
```
Comment (mark-lunarg) [inline layers/core_validation.cpp:1735]: Can you please share a short explanation of what's going on here?

File: layers/core_validation.h [hunk 3]
```diff
@@ -1441,6 +1441,24 @@ class CoreChecks : public ValidationStateTracker {
     bool PreCallValidateCmdSetFragmentShadingRateKHR(VkCommandBuffer commandBuffer, const VkExtent2D* pFragmentSize,
                                                      const VkFragmentShadingRateCombinerOpKHR combinerOps[2]) const override;
 
+
+    bool PreCallValidateCreateIndirectCommandsLayoutNV(VkDevice device, const VkIndirectCommandsLayoutCreateInfoNV* pCreateInfo,
```
Comment (mark-lunarg) [inline layers/core_validation.h:1445]: I think all of these guys (the Pre/PostCallValidate/Record declarations) will now need 'override' keywords -- at least one of the compilers will throw errors for this.

File: layers/core_validation_types.h [hunk 4]
```diff
@@ -1008,8 +1034,12 @@ class PIPELINE_STATE : public BASE_NODE {
         raytracingPipelineCI.initialize(&emptyRayTracingCI);
         stage_state.clear();
         fragmentShader_writable_output_location_list.clear();
+        shader_groups.clear();
+        shader_group_references.clear();
+        referenced_pipelines.clear();
     }
 
+    static void initVertexState(VertexState &vertex_state, const VkPipelineVertexInputStateCreateInfo *pVertexInputStateCreateInfo);
```
Comment (mark-lunarg) [inline layers/core_validation_types.h:1042]: I know the context is incorrect here (sorry!), would you mind changing these function names to UpperCamelCase please?

File: layers/parameter_validation_utils.cpp [hunk 5]
```diff
@@ -5391,6 +5494,342 @@ bool StatelessValidation::manual_PreCallValidateImportSemaphoreFdKHR(
...
+    }
+    if (pCreateInfo->tokenCount == 0 ||
+        pCreateInfo->tokenCount > phys_dev_ext_props.device_generated_cmds_props.maxIndirectCommandsTokenCount) {
+        skip |= LogError(device, "VUID-VkIndirectCommandsLayoutCreateInfoNV-tokenCount-02931",
+                         "vkCreateIndirectCommandsLayoutNV: tokenCount must be at least 1 and at maximum "
+                         "VkPhysicalDeviceDeviceGeneratedCommandsPropertiesNV::maxIndirectCommandsTokenCount.");
+    }
+
+    uint32_t stream_misaligned = 0;
+    bool hadWork = false;
```
Comment (mark-lunarg) [inline layers/parameter_validation_utils.cpp:5656]: Local var name reminder.

File: scripts/parameter_validation_generator.py [hunk 6]
```diff
@@ -240,6 +240,12 @@ def __init__(self,
             'vkGetAccelerationStructureBuildSizesKHR',
             'vkCmdWriteAccelerationStructuresPropertiesNV',
             'vkCreateDisplayModeKHR'
```
Comment (mark-lunarg) [inline scripts/parameter_validation_generator.py:242]: Missing a comma at the end of this line likely causing the manual routine calls to be skipped for the CreateDisplayModeKHR and CmdBindPipelineShaderGroupNV APIs.

**Comment by mark-lunarg** [comment]
@pixeljetstream, a ton of work here!  The content all looked fine, just some format nits and a couple of minor things.

File: layers/core_validation_types.h (hunk 4 above)
Comment (pixeljetstream) [reply layers/core_validation_types.h:1042]: all class functions?

File: scripts/parameter_validation_generator.py (hunk 6 above)
Comment (pixeljetstream) [reply scripts/parameter_validation_generator.py]: thx

File: layers/core_validation.cpp (hunk 2 above)
Comment (pixeljetstream) [reply layers/core_validation.cpp]: rephrased the comment with latest commit.

```
// When referenced pipelines are used, they must match in their creation state, except for 
// state that is overridden through shader groups and a few exceptions. This requires a deep 
// comparison of all known extension structs that the graphics state create info may
// reference.
//
// FIXME As of time of writing only basic structs are compared and therefore validation will not
// catch a state mismatch for extension structs. A full implementation would need to add 
// comparators for every extension struct, and find the corresponding struct in both
// pipelines and ensure a match.
// Following structs are known to be missing (as of header version 135)
```

Comment (pixeljetstream) [reply layers/core_validation.cpp]: not sure if this is good enough, a colleague may pick up on this work and add more infrastructure

File: layers/core_validation.h (hunk 3 above)
Comment (pixeljetstream) [reply layers/core_validation.h:1445]: done

**Comment by pixeljetstream** [comment]
> @pixeljetstream, a ton of work here! The content all looked fine, just some format nits and a couple of minor things.

thanks, yet to do a bit more testing on our samples, would it be okay to ship despite some functionality missing (see FIXME)/requiring unit tests? Or should it stay as separate branch until everything is done?

File: layers/core_validation.h [hunk 7]
```diff
@@ -266,7 +266,7 @@ class CoreChecks : public ValidationStateTracker {
     bool ValidateCmdDrawInstance(VkCommandBuffer commandBuffer, uint32_t instanceCount, uint32_t firstInstance, CMD_TYPE cmd_type,
                                  const char* caller) const;
     bool ValidateCmdDrawType(VkCommandBuffer cmd_buffer, bool indexed, VkPipelineBindPoint bind_point, CMD_TYPE cmd_type,
-                             const char* caller, VkQueueFlags queue_flags) const;
+                             const char* caller, VkQueueFlags queue_flags, const void* cmd_custom_info = nullptr) const;
```
Comment (pixeljetstream) [inline layers/core_validation.h:269]: @mark-lunarg while doing more implementation fixes I ran into an issue that I got the ordering for state_tracker and core_caldation calls wrong and assumed CB_STATE had been updated when it wasn't. As result I have to pass some state to these functions. Given that CMD_TYPE can be different, I chose a const void* here so call specific payloads can be passed. Let me know if that is okay.

File: scripts/layer_chassis_dispatch_generator.py [hunk 8]
```diff
@@ -195,6 +195,7 @@ class LayerChassisDispatchOutputGenerator(OutputGenerator):
             }
 
             local_pCreateInfos[idx0].initialize(&pCreateInfos[idx0], uses_color_attachment, uses_depthstencil_attachment);
+            WrapPnextChainHandles(layer_data, local_pCreateInfos[idx0].pNext);
```
Comment (pixeljetstream) [inline scripts/layer_chassis_dispatch_generator.py:198]: @mark-lunarg This is another bigger change with the latest fix, this seems to have been a generally missing before, but please have a look if I got it right.

File: layers/core_validation.h (hunk 7 above)
Comment (mark-lunarg) [reply layers/core_validation.h:269]: @pixeljetstream, can you point me to the spots where `cmd_custom_info `is getting cast to something other than `VkGeneratedCommandsInfoNV`?    Maybe also a more specific name too, like nv_custom_cmds_info or something like that?  The default-parameter approach will also need to be vetted by the new team.

File: layers/generated/object_tracker.cpp [hunk 9]
```diff
@@ -1054,51 +1054,6 @@ bool ObjectLifetimes::PreCallValidateMergePipelineCaches(
     return skip;
 }
 
-bool ObjectLifetimes::PreCallValidateCreateGraphicsPipelines(
```
Comment (mark-lunarg) [inline layers/generated/object_tracker.cpp:1057]: It'd really be best to keep this api completely generated, we'll have to see how the others feel about moving it to a manual function.

File: layers/core_validation.h (hunk 7 above)
Comment (pixeljetstream) [reply layers/core_validation.h:269]: there is nothing yet, I just figured other new drawcalls could use this, hence didn't want to imply it only has to be the nv generated cmds extension. If you think it should be specific until someone needs it to be void I can change type and name

Comment (mark-lunarg) [reply layers/core_validation.h:269]: Ah, I see.  Yeah, it's curious this hasn't come up before now.  I think if it was more than one thing, we'd likely add another struct as a new parameter and plumb it all the way down.  The other guys may want to do that anyway, so as I said, maybe we should just do the rebase and then run it by the rest of the team.

**Comment by juan-lunarg** [comment]
https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/4971

**Pull request closed without merging**